### PR TITLE
Issue 187

### LINKED ISSUE #187: Normalize DrugBank/other source concept IDs that omit namespaces
Eg, for search term `DB04050`, since it can be matched against a simple regex we should be able to try it as `drugbank:DB04050` if it otherwise fails to normalize. This is something we already perform within the `search/` endpoint and may be relevant to other normalizers as well.

### PR DESCRIPTION
**Pull request by jsstevenson**
Close #187 

* Update namespace inference mechanism. Previously, we defined prefix characters (like `DB` for DrugBank) and checked if an ID string started with them in order to infer a namespace. Now, we use LUI regular expressions as provided by identifiers.org. This is a little more accurate and also lets us capture ChemIDplus values too.
* Add namespace inference parameter, `infer`, to both `/search` and `/normalize` endpoints. Previously, this was done automatically in `/search` and wasn't available for `/normalize`. If true, tries to infer namespace as described above and provides a few details under `warnings`, including the inferred namespace value, the new query (I'm not sure this is necessary?), and `alternate_inferred_matches` (which may also be unnecessary because none of the regexes are compatible with each other).

Open questions:
* We don't retain LUI regexes that are too broad to be particularly specific (eg RxNorm IDs are just a series of numbers in the 1-7 digit range). Should we? This would make it more likely that we have alternate possible inferred matches, which opens up a few additional questions.
* Since we dropped NDA/ANDA from Drugs@FDA LUIs, we can't return responses for a value like "NDA017355" (only for "drugsatfda.nda:017355"). We could, however, use a regex like `^A?NDA\d+$` to recognize that type of string, and then perform a substitution to transform it into "drugsatfda.nda:017355". Would this be worthwhile?

### CODE REVIEW

**Comment by korikuzma** [comment]
> We don't retain LUI regexes that are too broad to be particularly specific (eg RxNorm IDs are just a series of numbers in the 1-7 digit range). Should we? This would make it more likely that we have alternate possible inferred matches, which opens up a few additional questions.

I think if the `infer` param is set, then we should try prefixing it with all source name prefixes to try and find a match.

> Since we dropped NDA/ANDA from Drugs@FDA LUIs, we can't return responses for a value like "NDA017355" (only for "drugsatfda.nda:017355"). We could, however, use a regex like ^A?NDA\d+$ to recognize that type of string, and then perform a substitution to transform it into "drugsatfda.nda:017355". Would this be worthwhile?

This approach makes sense and I think it should be added.

**Comment by jsstevenson** [comment]
> This approach makes sense and I think it should be added.

👍 Just pushed:

```
{
  "query": "nda018057",
  "warnings": [
    {
      "inferred_namespace": "drugsatfda.nda",
      "adjusted_query": "drugsatfda.nda:018057",
      "alternate_inferred_matches": []
    }
  ],
  "source_matches": [
    {
      "source": "DrugsAtFDA",
      "match_type": 100,
    ...
    }
  ]
}
```

**Review by mcannon068nw — DISMISSED** [review]

File: therapy/etl/wikidata.py [hunk 1]
```diff
@@ -28,6 +27,13 @@
     "wikidata": NamespacePrefix.WIKIDATA.value,
 }
 
+# Provide standard concept ID prefixes
+ID_PREFIXES = {
+    "wikidata": "Q",
+    "chembl": "CHEMBL",
+    "drugbank": "DB",
+    "ncit": "C",
+}
```
Comment (korikuzma) [inline therapy/etl/wikidata.py:36]: This is in `SourceIDAfterNamespace`. Maybe use it instead?

File: therapy/query.py [hunk 2]
```diff
@@ -442,15 +484,38 @@ def _handle_failed_merge_ref(self, record: Dict, response: Dict,
...
         return response
 
-    def search_groups(self, query: str) -> Dict:
+    def _resolve_merge(self, response: Dict, query: str,
+                       record: Dict, match_type: MatchType) -> Dict:
+        """Given a record, return the corresponding normalized record
+
+        :param Dict response: in-progress response object
+        :param str query: exact query as provided by user
+        :param Dict record: TODO
```
Comment (korikuzma) [inline therapy/query.py:493]: Update todo

File: therapy/main.py [hunk 3]
```diff
@@ -98,14 +104,19 @@ def search(q: str = Query(..., description=q_descr),
          response_description=merged_response_descr,
          response_model=NormalizationService,
          description=normalize_description)
-def normalize(q: str = Query(..., description=merged_q_descr)) -> Dict:
-    """Normalize query or concept ID provided by user. Return strongest-match merged
-    concept.
+def normalize(q: str = Query(..., description=merged_q_descr),
+              infer_namespace: Optional[bool] = Query(
```
Comment (korikuzma) [inline therapy/main.py:108]: essentially the same thing
```suggestion
              infer_namespace: bool = Query(
```

File: therapy/query.py [hunk 4]
```diff
@@ -442,15 +484,38 @@ def _handle_failed_merge_ref(self, record: Dict, response: Dict,
...
         """Return merged, normalized concept for given search term.
 
         :param str query: string to search against
+        :param bool infer: if true, try to infer namespace for IDs
         """
         # prepare basic response
         response: Dict[str, Any] = {
             "query": query,
-            "warnings": self._emit_warnings(query),
+            "warnings": self._emit_char_warnings(query),
```
Comment (korikuzma) [inline therapy/query.py:518]: This wasnt changed in this PR, but you don't need to set the url in ServiceMeta. Can change here or in a different issue

**Review by korikuzma — DISMISSED** [review]

File: therapy/etl/wikidata.py (hunk 1 above)
Comment (jsstevenson) [reply therapy/etl/wikidata.py:36]: Oh, I meant to pose this as a question and forgot -- `SourceIDAfterNamespace` doesn't get used anywhere outside of this line anymore, so I was thinking about moving it out of schemas and into here (like so). What're your thoughts?

File: therapy/main.py (hunk 3 above)
Comment (jsstevenson) [reply therapy/main.py:108]: Oh this is very true

File: therapy/etl/wikidata.py (hunk 1 above)
Comment (korikuzma) [reply therapy/etl/wikidata.py:36]: Ahh. If this is the only place we use it, we can leave it here for now. If we end up using it in other places, then we can move it to schemas then.

File: therapy/query.py (hunk 4 above)
Comment (jsstevenson) [reply therapy/query.py:518]: e z fix

**Review by korikuzma — APPROVED** [review]

**Pull request merged**